Source: Upstream
Upstream: http://bugs.mysql.com/bug.php?id=45058
Reason: init_available_charsets uses double checked locking.

diff -ur mysql-5.1.44.orig/include/my_sys.h mysql-5.1.44/include/my_sys.h
--- mysql-5.1.44.orig/include/my_sys.h	2010-03-15 19:16:52.000000000 +0100
+++ mysql-5.1.44/include/my_sys.h	2010-03-15 19:17:00.000000000 +0100
@@ -950,7 +950,7 @@
 extern my_bool resolve_collation(const char *cl_name,
                                  CHARSET_INFO *default_cl,
                                  CHARSET_INFO **cl);
-
+extern void free_charsets(void);
 extern char *get_charsets_dir(char *buf);
 extern my_bool my_charset_same(CHARSET_INFO *cs1, CHARSET_INFO *cs2);
 extern my_bool init_compiled_charsets(myf flags);
diff -ur mysql-5.1.44.orig/libmysql/libmysql.c mysql-5.1.44/libmysql/libmysql.c
--- mysql-5.1.44.orig/libmysql/libmysql.c	2010-03-15 19:16:52.000000000 +0100
+++ mysql-5.1.44/libmysql/libmysql.c	2010-03-15 19:17:00.000000000 +0100
@@ -211,6 +211,7 @@
   }
   else
   {
+    free_charsets();
     mysql_thread_end();
   }
 
diff -ur mysql-5.1.44.orig/mysys/charset.c mysql-5.1.44/mysys/charset.c
--- mysql-5.1.44.orig/mysys/charset.c	2010-03-15 19:16:52.000000000 +0100
+++ mysql-5.1.44/mysys/charset.c	2010-03-15 19:17:00.000000000 +0100
@@ -427,6 +427,11 @@
 }
 
 
+void free_charsets(void)
+{
+  charsets_initialized= MY_PTHREAD_ONCE_INIT;
+}
+
 uint get_collation_number(const char *name)
 {
   my_pthread_once(&charsets_initialized, init_available_charsets);
diff -ur mysql-5.1.44.orig/mysys/my_init.c mysql-5.1.44/mysys/my_init.c
--- mysql-5.1.44.orig/mysys/my_init.c	2010-03-15 19:16:52.000000000 +0100
+++ mysql-5.1.44/mysys/my_init.c	2010-03-15 19:17:00.000000000 +0100
@@ -165,6 +165,7 @@
       my_print_open_files();
     }
   }
+  free_charsets();
   my_error_unregister_all();
   my_once_free();
 
diff -ur mysql-5.1.44.orig/sql/mysqld.cc mysql-5.1.44/sql/mysqld.cc
--- mysql-5.1.44.orig/sql/mysqld.cc	2010-03-15 19:16:52.000000000 +0100
+++ mysql-5.1.44/sql/mysqld.cc	2010-03-15 19:17:00.000000000 +0100
@@ -1287,6 +1287,7 @@
   lex_free();				/* Free some memory */
   item_create_cleanup();
   set_var_free();
+  free_charsets();
   if (!opt_noacl)
   {
 #ifdef HAVE_DLOPEN
